PX4/Firmware
# Airspeed validation & selection module

This PR re-purposes the current wind estimator module into an airspeed (selection) module. 

**Overview**
- performs validity checks on airspeed sensor measurements
- if no valid airspeed sensor is available, it estimates airspeed by calculating groundspeed minus windspeed
- can estimate an airspeed scale factor and outputs an equivalent airspeed (EAS: indicated * scale ) (plus this scale can also be set via a parameter)
- can also handle multiple airspeed sensors (currently not yet supported from a drivers side) and switches to valid sensor if failure on one is detected

The airspeed valid checks are taken from https://github.com/PX4/Firmware/pull/11846 . They are here performed for every sensor and the code is located in the new airspeed selector module (currently the checks in commander are still there, but the idea is to only have them in the airspeed module in the end).

The output of the airspeed module is a airspeed validated topic, that gets logged but **currently doesn't affect any control modules**.  All control modules still subscribe to the "old" airspeed topic. Of course this is supposed to be changed in the future (fuse all airspeed data through this module before using it).

Here's a rough drawing: 
![image](https://user-images.githubusercontent.com/26798987/60331392-e18f8680-9994-11e9-84ba-0c350184ea96.png)

**Description**

The output of the module is a airspeed_validated topic, containing:
- indicated airspeed (IAS)
- equivalent airspeed (EAS)
- true airspeed (TAS)
- true_ground_minus_wind_m_s airspeed estimate
- indicated_ground_minus_wind_m_s airspeed estimate
- index of selected airspeed sensor (0: first sensor, 1: second sensor, -1: groundspeed-windspeed, -2: invalid airspeed)

For failure detection as well as for the estimation of a scale factor from IAS to EAS, per airspeed sensor a wind estimator is run, fusing sideslip and the airspeed measurement from the corresponding sensor. Another wind estimator is run to estimate the windspeed used for the groundspeed-windspeed airspeed estimate (this wind estimator is only fusing sideslip, no airspeed measurement).

Detailed diagram of new Module:
https://www.draw.io/#G1xrT_nPgBEt5Qh5dGYbX-CyKz9cROWnRT

**Usage**
1. Performing airspeed checks
The same/very similar checks as with the current checks in commander. Also takes the same parameter to set them (eg COM_TAS_FS_INNOV for innovation threshold). If the airspeed measurement for the sensor is declared invalid, the module outputs all airspeed(IAS, EAS, TAS) based on windpeed-groundspeed (if local position is valid) and sets the flag _airspeed_sensor_measurement_valid_ to false. If local position is also invalid, then _airspeed_valid_ is set to false. 

2. Estimating airspeed scale
In-flight estimation of the scale factor (EAS = scale * IAS) is turned on by setting the parameter ARSP_SCALE_EST to 1. Then the factor is estimated inside the wind estimator (otherwise it is set to 1 inside the wind estimator). After some minutes of flight, the scale should have converged to the correct value, and the estimation can be switched off by setting ARSP_SCALE_EST to 0. By doing this, the estimated scale is written to the parameter ARSP_ARSP_SCALE and keeps being used to scale the airspeed. The value is limited to 0.75..1.25. 
The scale can also be set manually by setting ARSP_ARSP_SCALE manually.

**Test data / coverage**
1. SITL (VTOL): Procedure:
- after 57s, ARSPV_SCALE_EST is set to 1 (starting estimation of scale factor)
- at 80s, ARSPV_SCALE_EST is set to 0 (ending estimation). Now the estimated value is printed in log (see below) and the parameter ARSPV_ARSP_SCALE is set to the estimated value (in this case 1, as in SITL we don't have sensor placement/blockage errors)
- at 110s, the stall speed of the vehicle (COM_ASPD_STALL) is set much too high (50), thus triggering the load factor check (flying below stall speed). The airspeed sensor is detected as invalid, and the index switches to -1 (groundspeed-windspeed). 
logs: https://review.px4.io/plot_app?log=e63b36b6-53d1-48ef-aaea-6151ad5487cd
Messages in QGC:
![airspeed_selector_sitl_log](https://user-images.githubusercontent.com/26798987/61041620-ee6b9b80-a3d2-11e9-879a-d08ac72a3d79.png)


2. Flight test log: scale estimation, load factor and innovation check fail
Procedure:
- enable scale estimation at 57s, disable at 119s --> estimated scale 1.03
- at 150s, set stall speed to 30, triggers load factor fail --> switched to ground-windspeed
- reset stallspeed to 6, airspeed declared valid again
- repeat the the load factor triggering while doing automatic loitering (ca at 170s and 200s)
- set innovation check levels (COM_TAS_FS_INNOV and COM_TAS_FS_INTG) to 0.1 -->innovation check failure is triggered as windspeed innovation are above these limits --> goes to ground-windspeed
-set the innovation check params to normal values --> airspeed is declared valid again

![image](https://user-images.githubusercontent.com/26798987/61059697-bd529180-a3f9-11e9-9996-f3682ae1325d.png)
In these plots you can see how the airspeed_validated/true_airspeed is bounded between 12.5 and 14.2m/s while the artificial airspeed check failures are produced and it is thus sometimes calculated from groundspee-windspeed (when measurement not valid) or from pitot measurement with scale applied.  

We also did the scale estimation while loitering (not in this flight), with much slower convergence. Should consider tuning the wind estimator to make if faster.  

https://review.px4.io/plot_app?log=1c98f1a2-6d0e-4a79-b065-2f67b603228f

3. Bench test: Detection of data_missing check:
(for this test we connected 2 airspeed sensors and made each publish an airspeed topic)
![image](https://user-images.githubusercontent.com/26798987/60287021-2d98e780-9911-11e9-988d-760f078fbf27.png)
Description: data_missing failure was triggered by unplugging the airspeed sensor with index 0. The airspeedValidator detects that and switches to index 1 (second sensor). Next all the airspeed sensors are plugged out, and the validator switches to index -2 (indicating no valid airspeed, as groundspeed-windspeed is not valid as performed indoor without GPS). Afterwards, the sensors are reconnected and the valid sensor index goes back to 0. 

**Further work**
As mentioned, this PR only brings in the new module passively, without effect on any control modules. The idea is to get feedback and work e.g. on the tuning of the validity checks, and also to get it tested with higher winds (so far the max we tested it with was about 4 m/s). Follow up work is:
- enable multiple airspeed sensors from a driver side (also consider calibration)
- get rid of checks in commander and adapt commander to consider the airspeed_validated topic, and take action based on the validity flags of this topic (e.g. RTL when airspeed is invalid)
- use the airspeed_validated topic in control modules (e.g. EAS replaces the old IAS for attitude controller, TAS from the validated topic for TECS/L1)
